ElectricLoadCenter:Storage:Simple Zero Efficiency Fix

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8867 
 - A user noticed some strange "infinite" results in the output during random times.  The file included ElectricLoadCenter:Storage:Simple and a simple PV generator.  The problem was traced back to the fact that the user's input file had a zero efficiency for both charging and discharging and that there were several places in the code where those parameters were in the denominator of an equation resulting in a NaN calculation.  While EnergyPlus did not crash, the results did not make sense.  Changing the values to non-zero in the input corrected the problem, but that is just a work-around.  This fix changed the minimum input to 0.001 for both of the efficiencies and also included code changes to avoid any divide by zero possibility even if somehow a 0 value gets through the Input Processor.  Documentation changes and a unit test are included in this work.  No changes to the output of the program resulted from this work, as expected.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correct Divide by Zero in Simple Electric Storage

Corrected the code and the IDD to avoid a divide by zero in the simple (bucket) storage model of the electric load center.  Before the fix, a zero value was allowed for both the charging and discharging efficiencies.  These both led to a divide by zero as shown in a user file.  This has been corrected.

RKStrand committed:
Unit Test for Simple Electric Storage

A unit test and another code correction to avoid any possibility of a divide by zero.  This way, even if someone changes the minimum in the IDD back to 0, a divide by zero will be avoided.

jcyuan2020 commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3944,6 +3944,8 @@ void ElectricStorage::simulateSimpleBucketModel(EnergyPlusData &state,
                                                 Real64 const controlSOCMinFracLimit)
 {
+    Real64 const minEfficiency = 0.001;
+
```
An optional suggestion: the implementation assumes a common `minEfficiency' for both the charge and the discharge efficiency values. It might have better clarity by using two variables like minChrgEfficiency and minDischrgEfficiency ( and use them at their own corresponding code lines below)---and also so that the two min settings can be tuned independently in the future if needed.

RKStrand replied:
See below--I have followed your suggestion and simplified the code in the process.  There are now separate efficiency parameters for charging and discharging and the actual values are simply overwritten when the efficiencies are less than the minimum.  There is now a single equation for each case rather than 2.  Hopefully the ci comes back clean and it will be ready for re-review soon.

mjwitte replied:
So, this will just silently set efficiency to 0.001? Seems like that should either be a severe error (and require a different input) or at least warn that it's been changed.

RKStrand replied:
The limits on the input should prevent this from happening, but I will put in a warning message here that the value has been reset.

mjwitte replied:
Wait, is this being applied during the simulation? Maybe I'm misunderstanding what's happening here. I thought this was a check on a simple input.

RKStrand replied:
Warning messages have now been added for both efficiencies.

jcyuan2020 commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3998,8 +4007,13 @@ void ElectricStorage::simulateSimpleBucketModel(EnergyPlusData &state,
+            thisTimeStepStateOfCharge_ = lastTimeStepStateOfCharge_ - powerDischarge * state.dataHVACGlobal->TimeStepSys *
+                                                                          DataGlobalConstants::SecInHour / energeticEfficDischarge_;
+        } else {
+            thisTimeStepStateOfCharge_ =
```
For this condition, can we just override the `energeticEfficDischarge` with the minimum efficiency. And then one common formula as shown in Lines 4011-4012 can be used?

RKStrand replied:
Done.  Just committed changes to address this comment.  The code now simply overrides as requested.

jcyuan2020 commented on src/EnergyPlus/ElectricPowerServiceManager.cc:
```diff
@@ -3977,10 +3984,12 @@ void ElectricStorage::simulateSimpleBucketModel(EnergyPlusData &state,
-                                              energeticEfficDischarge_) <= (maxEnergyCapacity_ * controlSOCMinFracLimit)) {
-            powerDischarge = (lastTimeStepStateOfCharge_ - (maxEnergyCapacity_ * controlSOCMinFracLimit)) * energeticEfficDischarge_ /
-                             (state.dataHVACGlobal->TimeStepSys * DataGlobalConstants::SecInHour);
+        if (energeticEfficDischarge_ >= minEfficiency) {
```
Also, here can be override the `energeticEfficDischarge` in earlier stages so that here it would be guaranteed to be greater than or equal to the minimum efficiency; then it can share the same equation?

RKStrand replied:
Done. Just committed changes to address this comment. The code now simply overrides as requested.

jcyuan2020 commented on idd/Energy+.idd.in:
```diff
@@ -88610,10 +88610,10 @@ ElectricLoadCenter:Storage:Simple,
        \maximum 1.0
-       \minimum 0.0
+       \minimum 0.001
   N4,  \field Maximum Storage Capacity
```
Great!

jcyuan2020 commented on tst/EnergyPlus/unit/ElectricPowerServiceManager.unit.cc:
```diff
@@ -1174,4 +1174,110 @@ TEST_F(EnergyPlusFixture, Battery_LiIonNmc_Simulate)
+    ExpectedResult = 2800.0;
+    battery.simulate(*state, powerCharge, powerDischarge, charging, discharging, controlSOCMaxFracLimit, controlSOCMinFracLimit);
+    EXPECT_NEAR(powerDischarge, ExpectedResult, 0.01);
+}
```
Nice unit test!

jcyuan2020 reviewed (commented):
I think with the root cause has been successfully identified and the work would nice fix the reported issue. Some minor suggestions might be considering to split the two minimum efficiency values (one charge and one discharge0 in the implementation. Also wondering if the actual efficiency variables (`energeticEfficCharge_` and `energeticEfficDischarge_`) could just be overridden to the minimum allowed value (0.001) if a lower-than-minimum value is inputted, which might be able to slightly simplify the scenarios.

RKStrand committed:
Electric Load Error Review Comments

Changes to reflect the review comments for this defect.  A single parameter was made into two separate parameters and the code was significantly simplified so the problematic variables simply get reset to the minimum if they are ever less than the limit.

RKStrand committed:
Changing Efficiency Minimum Fix

Fix to an obvious error in the logic of the efficiency parameter assignment.

jcyuan2020 reviewed (approved):
I ran the defect file with the current branch. The simulation runs successfully when the input is correctly inputted an efficient value of at least 0.001. And also, if the two efficiency values are accidentally left blank (as the original defect file did), the following error messages would prompt up to warn the user to fix the inputs: 
```
   ** Severe  ** <root>[ElectricLoadCenter:Storage:Simple][BatterySimple][nominal_discharging_energetic_efficiency] - "0" - Expected number greater than or equal to 0.001000
   ** Severe  ** <root>[ElectricLoadCenter:Storage:Simple][BatterySimple][nominal_energetic_efficiency_for_charging] - "0" - Expected number greater than or equal to 0.001000
   **  Fatal  ** Errors occurred on processing input file. Preceding condition(s) cause termination.
```
This is nice, and I think the error warnings definitely have some significance in reminding the user: it is obviously not reasonable in the original defect file that the user left the two efficiency fields blank. Even if the program resets these to 0.001 minimum to run OK, it could still be a somehow illly configured case (I assume that 0.001 is quite low for a healthy running equipment?). So here the error messages served as an important remind to the user that these efficiency values need to be changed to some values in range (0.001 to 1) or something more realistic. 

An additional thought might be supplying some reasonable default to be set to these values in the IDD? I see that the three example cases use 0.8 (explicitly inputted, not default)---I am not sure if this 0.8 is adequate as a default, or even if it is reasonable to set a default value at all for these fields--so feel free to give some thoughts or suggestions. But anyway, this should not hinder the current branch to get merged as is.

RKStrand commented:
@jcyuan2020 Yes, 0.001 is pretty low, but I am not sure what a default should be.  0.7?  0.8?  Higher?  It would be nice to have someone investigate this and come up with a default, but I agree that this is not part of this defect which was to fix the NaN problem.  Thanks for your review!

mjwitte commented:
@RKStrand Please merge in develop for final review.

RKStrand committed:
Merge branch 'develop' into 8867ElectricLoadCenterSimpleStorageDivideByZero

RKStrand commented:
@mjwitte No problem--just did that now.  Hopefully it all comes back good.

mjwitte commented:
@RKStrand Just noticed that this has an idd change. That will have to wait until after 9.6 release.

RKStrand committed:
Electric Power Storage min Efficiency Warning Messages

Warning messages were added to the code to make sure the users know if the minimum efficiencies were rest to avoid a divide by zero.  This was added to address review comments.

RKStrand commented:
@mjwitte Oh you are right.  I forgot to flag this prior to the IO Freeze because I forgot about that detail.  Any change we can move forward with the rest of it (everything else minus the minor IDD change) in the meantime?  The checks in the code will still prevent the problem from happening.  Basically, if the user puts in a zero efficiency, they will get a divide by zero.  All this is trying to do is avoid that by making sure that the efficiencies are greater than zero.

RKStrand committed:
Electric Load Center Walkback of IDD Change

We are past the IO Freeze date.  This was turned into a PR before the deadline, but it was not properly flagged to be looked at prior to the IO Freeze deadline.  The code changes are still good though.

RKStrand commented:
@mjwitte Walked back the IDD change and pushed a new version back up.  Now, there should be no IDD changes.  Hopefully ci comes back clean and this is good to go.  You okay with the IDD getting added after the release?

mjwitte commented:
@RKStrand I'm ok with delaying the IDD change, but I'm not ok with checking min efficiency (and throwing warnings) on every call to `simulateSimpleBucketModel`. Am I misunderstanding this?

RKStrand commented:
@mjwitte It should only do that once, right?  Those parameters are inputs that never get reset.  If they come in and get reset by the if statement, it shouldn't ever hit those warnings again because it has been reset to something that won't trip the if.  So, the if conditions should only hit once and not every time it comes in.  At least I think that's the case.  Did I miss something here?

mjwitte commented:
Well, true, but the if's are needlessly executed on every call. Can this be moved into the getinput function?

RKStrand commented:
@mjwitte Ah, yes, good point.  I'll move that and come up with a different unit test.  Shouldn't take too long and will definitely be a better solution than having it in the simulate routine.  Surprised I didn't think of that when I worked on this...

RKStrand committed:
Electric Storage Review Comments

This commit addresses the review comments that were concerned with the placement of the resetting of the efficiencies.  This now happens once at the getting of the input.  A new unit test for this revision was also created.

RKStrand committed:
Merge branch 'develop' into 8867ElectricLoadCenterSimpleStorageDivideByZero

RKStrand commented:
@mjwitte Ok, rewrote this.  Checks for minimum levels of efficiency have been moved so that they are done during the getting of the input.  So, this should only happen once and not clutter up/slow down the simulate portion of the routine.  New unit test to reflect this change.  Commit made and merged in current develop.  Hopefully this is better now.

RKStrand committed:
Merge branch 'develop' into 8867ElectricLoadCenterSimpleStorageDivideByZero

mjwitte reviewed (commented):
@RKStrand Ran the defect file and this still has a divide by zero. It's not just the user input that's a problem. `DCtoACInverter::calcEfficiency` is setting `efficiency_` to zero at line 2884 when plr is zero. `powerOutOfInverter` is zero, so this is doing zero/zero.

RKStrand committed:
8867 Correction to Another Divide by Zero

During review, another location (or two) where a divide by zero might result from an efficiency_ being set to 0.  These are cases when the load is zero, so setting the efficiency_ to 1 should not impact anything and avoid the divide by zero problems.

RKStrand committed:
Merge branch 'develop' into 8867ElectricLoadCenterSimpleStorageDivideByZero

RKStrand commented:
@mjwitte Ok, I've fixed that and one other occurrence of an "efficiency_ = 0.0;" line.  It's now committed (and merged in develop) and waiting on the ci to run.  Hopefully this takes care of the divide by zeroes.

jcyuan2020 commented:
I run a local CI on the few test files that showed the diffs in the web portal. The diffs were explained by the corrected efficiency (which occurs when plr is <=0 according to the code). The defect file finished running without problems.

jcyuan2020 commented:
One minor concern is that now the warning level seems to be somehow "downgraded". The very first idd-change version would trigger idd/schema check and will break the simulation with the warning about the two efficiency values. The current version is a regular warning and change to 0.001 and then simulation continues---the warning is kind of less ostensible than the original one, assuming that the user probably should consider inputting a more realistic value like 0.6-0.8 etc.

RKStrand commented:
I can bump up the level of the error message and make it a severe/fatal.  That would mimic what would happen if this was in the IDD.  I'll work on getting that in today.

RKStrand committed:
8867 Electric Load Center Address Review Comments

Addressed additional review comments.  Error message is now severe with fatal exit just like it would be for checks done in the IDD.

RKStrand committed:
Merge branch 'develop' into 8867ElectricLoadCenterSimpleStorageDivideByZero

RKStrand commented:
@jcyuan2020 Ok, I've modified this, committed the change, and merged develop into this branch.  CI is running.  Hopefully this is good to go now.  Thanks for reviewing!

jcyuan2020 commented:
@RKStrand Thanks! I did a local CI run on the few files that showed diffs on the portal, so hopefully no longer need to wait for the CI. The results look good and diffs are all expected. The warning level was now bumped up and will let the user to check the error to run again. All good now.

mjwitte commented:
Thanks @RKStrand and @jcyuan2020 merging.
